### PR TITLE
Remove dead code only if function name is replaced

### PR DESCRIPTION
**Pull request by reazulhoque**
This PR makes sure we try to remove dead code only when we replace the Numpy function names with our own. Removing dead code regardless was failing in cases where we have multiple `prange` as mentioned in [this issue](https://github.com/IntelPython/numba-dppy/issues/302). This PR should close https://github.com/IntelPython/numba-dppy/issues/302.

### CODE REVIEW

**Review by PokhodenkoSA — CHANGES_REQUESTED** [review]
Please, provide test case.

**Comment by reazulhoque** [comment]
@PokhodenkoSA this bug was noticed originally because [this test](https://github.com/IntelPython/numba-dppy/blob/main/numba_dppy/tests/test_prange.py) was falling back. This PR fixes this test from falling back and offloads it through our pipeline. Should we ask the original author of that test to make sure it fails if it is falling back instead of going through our pipeline?

I can always write tests and that would require setting up a partial compiler pipeline to test it, which is tedious. Please share your thoughts.

**Comment by PokhodenkoSA** [comment]
> [this test](https://github.com/IntelPython/numba-dppy/blob/main/numba_dppy/tests/test_prange.py) was falling back.

Does it mean that test did not check that code falling back? So test was blind?
No when you fix the code did you fix the test to check that fallback is not happening now?
I think test should be fixed first to catch bad situation and show that the test fails. Then your fix for the feature should show that it makes test passing.

**Comment by reazulhoque** [comment]
>> Does it mean that test did not check that code falling back? So test was blind?
Yes.

I have updated the tests to check for falling back.

File: numba_dppy/tests/test_prange.py [hunk 1]
```diff
@@ -81,9 +88,13 @@ def f(a, b):
...
         b = np.ones((m, n))
 
-        with dpctl.device_context("opencl:gpu"):
+        old_debug = numba_dppy.compiler.DEBUG
+        numba_dppy.compiler.DEBUG = 1
+
+        with captured_stdout() as stdout, dpctl.device_context("opencl:gpu"):
             f(a, b)
 
+        self.assertTrue("Parfor lowered to specified SYCL device" in stdout.getvalue())
```
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_prange.py:97]:
```suggestion
        self.assertIn("Parfor lowered to specified SYCL device", stdout.getvalue())
```

File: numba_dppy/tests/test_prange.py [hunk 2]
```diff
@@ -135,12 +150,14 @@ def prange_example():
         numba_dppy.compiler.DEBUG = old_debug
 
         self.assertEqual(
-            stdout.getvalue().count("Parfor lowered on DPPY-device"),
+            stdout.getvalue().count(""),
```
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_prange.py:153]:
```suggestion
            stdout.getvalue().count("Parfor lowered to specified SYCL device"),
```

File: numba_dppy/tests/test_prange.py [hunk 3]
```diff
@@ -37,7 +33,7 @@ def f(a, b):
         a = np.ones((m, n))
         b = np.ones((m, n))
 
-        with dpctl.device_context("opencl:gpu"):
+        with assert_auto_offloading(1, 0), dpctl.device_context("opencl:gpu"):
```
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_prange.py:36]:
```suggestion
        with assert_auto_offloading(parfors=1, 0), dpctl.device_context("opencl:gpu"):
```

File: numba_dppy/testing.py [hunk 4]
```diff
@@ -76,3 +77,36 @@ def assert_dpnp_implementaion():
         yield
 
     assert "dpnp implementation" in stdout.getvalue(), "dpnp implementation is not used"
+
+
+@contextlib.contextmanager
+def assert_auto_offloading(
+    expected_offloaded_count=-1, expected_not_offloaded_count=-1
```
Comment (PokhodenkoSA) [inline numba_dppy/testing.py:84]:
```suggestion
    parfors=-1, expected_not_offloaded_count=-1
```

Comment (PokhodenkoSA) [inline numba_dppy/testing.py:84]: We could find better name for parameters.

File: numba_dppy/tests/test_prange.py (hunk 3 above)
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_prange.py:36]: Use keyword arguments.

File: numba_dppy/testing.py (hunk 4 above)
Comment (PokhodenkoSA) [reply numba_dppy/testing.py:84]: Use 0 as default value for the second parameter.

File: numba_dppy/testing.py [hunk 5]
```diff
@@ -76,3 +77,36 @@ def assert_dpnp_implementaion():
...
+):
+    old_debug = numba_dppy.compiler.DEBUG
+    numba_dppy.compiler.DEBUG = 1
+
+    with captured_stdout() as stdout:
+        yield
+
```
Comment (PokhodenkoSA) [inline numba_dppy/testing.py:91]: Reset DEBUG to old value.

**Comment by reazulhoque** [comment]
@PokhodenkoSA @diptorupd Can you please review it?

**Review by PokhodenkoSA — APPROVED** [review]

**Pull request merged**